File: boligportal.dk/crawler.go
package main

import (
	"encoding/json"
	"io/ioutil"
	"net/http"
	"time"
)

// divide a DKK amount by this rate to get the price in euros
const DKKToEuroRate = 7.46635

type Location struct {
	Lat float64 `json:"lat"`
	Lon float64 `json:"lon"`
}

type Offer struct {
	Identifier  string   `json:"identifier"`
	Title       string   `json:"title"`
	Description string   `json:"description"`
	Price       int      `json:"price"`
	Currency    string   `json:"currency"`
	PriceEur    int      `json:"price_eur"`
	Area        int      `json:"area"`
	Rooms       int      `json:"rooms"`
	Location    Location `json:"location"`
}

func Crawl(url string) (offers []Offer, err error) {
	var httpClient = &http.Client{Timeout: 10 * time.Second}

	response, err := httpClient.Get(url)
	if err != nil {
		return offers, err
	}

	defer response.Body.Close()

	jsonPayload, err := ioutil.ReadAll(response.Body)
	if err != nil {
		return offers, err
	}

	var payload map[string]interface{}
	json.Unmarshal(jsonPayload, &payload)

	data := payload["data"].(map[string]interface{})
	properties := data["properties"].(map[string]interface{})
	propertiesCollection := properties["collection"].([]interface{})

	for _, propertyData := range propertiesCollection {
		offerData := propertyData.(map[string]interface{})

		price := int(offerData["monthlyPrice"].(float64))

		offer := Offer{
			Identifier:  "https://www.boligportal.dk/en" + offerData["url"].(string),
			Title:       offerData["title"].(string),
			Description: offerData["description"].(string),
			Price:       price,
			Currency:    "DKK",
			PriceEur:    int(float64(price) / DKKToEuroRate),
			Area:        int(offerData["sizeM2"].(float64)),
			Rooms:       int(offerData["numRooms"].(float64)),
			Location:    Location{Lat: offerData["lat"].(float64), Lon: offerData["lng"].(float64)},
		}

		offers = append(offers, offer)
	}

	return offers, err
}

File: leboncoin/crawler.py
#!/usr/bin/env python3

import http.client
import json

class Crawler:
    API_HOST = 'api.leboncoin.fr'
    API_URL = '/finder/search'
    API_KEY = 'ba0c2dad52b3ec'

    def offers(self):
        data = self._query()

        for offer_data in data['ads']:
            yield {
                'identifier': offer_data['url'],
                'title': offer_data['subject'],
                'description': offer_data['body'],
                'price': offer_data['price'][0],
                'currency': 'EUR',
                'price_eur': offer_data['price'][0],
                'location': {
                    'lat': offer_data['location']['lat'],
                    'lon': offer_data['location']['lng'],
                },
                'rooms': int(self._extract_attr(offer_data, 'rooms', 0)),
                'area': int(self._extract_attr(offer_data, 'square', 0)),
            }

    def _extract_attr(self, offer_data, attr_name, default = None):
        attribute_value = [attribute_data['value'] for attribute_data in offer_data['attributes'] if attribute_data['key'] == attr_name]

        try:
            return attribute_value[0]
        except IndexError:
            return default

    def _query(self):
        request_payload = '{"limit":35,"limit_alu":3,"filters":{"category":{"id":"10"},"enums":{"real_estate_type":["2"],"ad_type":["offer"]},"location":{"city_zipcodes":[{"city":"Lyon","label":"Lyon (toute la ville)","departement_id":69, "locationType": "city"}],"regions":["22"]}}}'
        headers = {"Content-type": "application/json", "User-Agent": "Mozilla/5.0 (X11; Linux x86_64; rv:61.0) Gecko/20100101 Firefox/61.0", "api_key": self.API_KEY}

        conn = http.client.HTTPSConnection(self.API_HOST)
        conn.request('POST', self.API_URL, request_payload, headers)
        response = conn.getresponse()

        if response.status != 200:
            raise Exception('Error while calling the API')

        return json.loads(response.read().decode('utf-8'))
